commit 2913219dda733a9c07e4d0837bfba983e46e80e2 (refs/changes/33/1033/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-12-01T06:14:06+00:00 (2 years, 10 months ago)
    
    CNS-181 limit size of IXFR messages
    Reviewed by: Richard Kiene <richard.kiene@joyent.com>

diff --git a/lib/dns-server.js b/lib/dns-server.js
index 4e6ade1..3b5b89f 100644
--- a/lib/dns-server.js
+++ b/lib/dns-server.js
@@ -31,6 +31,8 @@ var NS_TTL = consts.NS_TTL;
 var VERSION = consts.VERSION;
 var MAX_NOTIFY_FAILURES = 5;
 
+var MAX_ANS_PER_MSG = 500;
+
 function DNSServer(opts) {
 	assert.object(opts, 'options');
 
@@ -1069,7 +1071,7 @@ DNSServer.prototype.handleTransfer = function (q, z, cb) {
 					self.addAnswers(q, recs,
 					    name + '.' + z);
 					pushed += recs.length;
-					if (pushed >= 100) {
+					if (pushed > MAX_ANS_PER_MSG) {
 						q.send();
 						pushed = 0;
 					}
@@ -1089,7 +1091,7 @@ DNSServer.prototype.sendIxfrDiff = function (q, z, from, to, soa, cb) {
 
 	var i;
 	var nsDiff;
-	var pushed = 0;
+	var pushed = q.response.answer.length;
 	var pushRecs = function (recs) {
 		if (typeof (recs[0]) === 'string')
 			recs = recs.map(JSON.parse);
@@ -1100,7 +1102,7 @@ DNSServer.prototype.sendIxfrDiff = function (q, z, from, to, soa, cb) {
 			self.addAnswers(q, [recs[i].record],
 			    recs[i].name ? (recs[i].name + '.' + z) : z);
 			pushed++;
-			if (pushed > 100) {
+			if (pushed > MAX_ANS_PER_MSG) {
 				q.send();
 				pushed = 0;
 			}
@@ -1140,7 +1142,7 @@ DNSServer.prototype.sendIxfrDiff = function (q, z, from, to, soa, cb) {
 			soa.serial = from;
 			q.addAnswer(z, soa, TTL);
 			pushed++;
-			if (pushed > 100) {
+			if (pushed > MAX_ANS_PER_MSG) {
 				q.send();
 				pushed = 0;
 			}
@@ -1159,7 +1161,7 @@ DNSServer.prototype.sendIxfrDiff = function (q, z, from, to, soa, cb) {
 				soa.serial = to;
 				q.addAnswer(z, soa, TTL);
 				pushed++;
-				if (pushed > 100) {
+				if (pushed > MAX_ANS_PER_MSG) {
 					q.send();
 					pushed = 0;
 				}
